services/src/main/java/com/linkedin/kmf/services/TopicManagementService.java

```diff
@@ -0,0 +1,419 @@
+            topicState = topicState();
+          }
+          if (topicState.someBrokerMissingPartition()) {
+            LOG.info(_serviceName + ": rebalancing monitored topic.");
+            waitForTopicAssignmentsToComplete();
+            reassignPartitions(topicState._allBrokers, topicState._partitionInfo.size(), topicState.replicationFactor());
+            waitForTopicAssignmentsToComplete();
+            topicState = topicState();
+          }
+          if (topicState.someBrokerWithoutLeader()) {
+            LOG.info(_serviceName + ": running preferred replica election.");
+            triggerPreferredLeaderElection(_zkUtils, topicState._partitionInfo);
+          }
+          LOG.info(_serviceName + ": topic rebalance complete.");
+        } else {
+          LOG.info(_serviceName + ": topic is in good state, no rebalance needed.");
+        }
+      } catch (Exception e) {
+        if (e instanceof IOException) {
+          //Can't do this with catch block because nothing declares IOException although scala code can still throw it.
+          LOG.error(_serviceName + ": will retry later due to IOException.", e);
+        } else {
+          LOG.error(_serviceName + ": monitored topic rebalance failed with exception.  Exiting rebalance service.", e);
+          stop();
+        }
+      }
+    }
+  }
+
+  private final double _partitionBrokerRatioThreshold;
+  private final String _topic;
+  private final int _partitionsPerBroker;
+  private final int _scheduleIntervalMs;
+  private final ZkUtils _zkUtils;
+  private final ScheduledExecutorService _executor;
+  private volatile boolean _running = false;
+  private final String _serviceName;
+
+
+  public TopicManagementService(Map<String, Object> props, String serviceName) {
+    _serviceName = serviceName;
+    TopicManagementConfig config = new TopicManagementConfig(props);
+    _partitionBrokerRatioThreshold = config.getDouble(TopicManagementConfig.PARTITIONS_PER_BROKER_THRESHOLD);
+    _topic = config.getString(TopicManagementConfig.TOPIC_CONFIG);
+    _partitionsPerBroker = config.getInt(TopicManagementConfig.PARTITIONS_PER_BROKER_CONFIG);
+    _scheduleIntervalMs = config.getInt(TopicManagementConfig.REBALANCE_INTERVAL_MS_CONFIG);
+    String zkUrl = config.getString(TopicManagementConfig.ZOOKEEPER_CONNECT_CONFIG);
+    _zkUtils = ZkUtils.apply(zkUrl, ZK_SESSION_TIMEOUT_MS + _scheduleIntervalMs, ZK_CONNECTION_TIMEOUT_MS + _scheduleIntervalMs,
+      JaasUtils.isZkSecurityEnabled());
+    _executor = Executors.newScheduledThreadPool(1);
+
+    LOG.info("Topic management service \"" + _serviceName + "\" constructed with expected partition/broker ratio " + _partitionBrokerRatioThreshold
+      + " topic " + _topic + " partitionsPerBroker " + _partitionsPerBroker + " scheduleIntervalMs " +
+      _scheduleIntervalMs + ".");
+  }
+
+
+  /**
+   *  Polling...
+   */
+  private void waitForAddedPartitionsToBecomeActive(int expectedNumberOfPartitions) {
```

Comment:
It seems that this method is not needed because `AdminUtils.addPartitions(..)` will update the same znode that this method reads.